helper/iterator/iterator.go

```diff
@@ -0,0 +1,23 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package iterator
+
+// Iterator represents an object that can iterate over a set of values one at a
+// time.
+type Iterator interface {
+	// Next returns the next element or nil if there are none left.
+	Next() any
+}
+
+// Len consumes the iterator and returns the number of elements found.
+//
+// IMPORTANT: this method consumes the iterator, so it should not be used after
+// Len() returns.
+func Len(iter Iterator) int {
```

Comment:
At some point I'd love to see if we can get something like this done upstream in go-memdb (non-consuming if possible!), but it hasn't quite made it to the top of my todo pile :grinning:

Reply:
I could swear there was one already but I couldn't find it 😅 

Not consuming the iterator would be tricky. I don't think we can create a deep copy or reset the iterator somehow.

Reply:
Yeah I think you're right; the `ResultIterator` interface always bottoms out in the iradix library, but the results are filtered before they get to the caller so we couldn't use some count object on the radix trie to help us anyways. :shrug: